src/python/pants/backend/go/util_rules/cgo.py

```diff
@@ -0,0 +1,858 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import dataclasses
+import logging
+import os
+import shlex
+import textwrap
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Iterable
+
+from pants.backend.go.subsystems.golang import GolangSubsystem
+from pants.backend.go.util_rules import cgo_binaries, cgo_pkgconfig
+from pants.backend.go.util_rules.cgo_binaries import CGoBinaryPathRequest
+from pants.backend.go.util_rules.cgo_pkgconfig import (
+    CGoPkgConfigFlagsRequest,
+    CGoPkgConfigFlagsResult,
+)
+from pants.backend.go.util_rules.cgo_security import check_linker_flags
+from pants.backend.go.util_rules.goroot import GoRoot
+from pants.backend.go.util_rules.sdk import GoSdkProcess
+from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
+from pants.core.util_rules.system_binaries import (
+    BashBinary,
+    BashBinaryRequest,
+    BinaryPath,
+    BinaryPathTest,
+)
+from pants.engine.engine_aware import EngineAwareParameter
+from pants.engine.environment import Environment, EnvironmentRequest
+from pants.engine.fs import (
+    CreateDigest,
+    DigestContents,
+    DigestSubset,
+    Directory,
+    FileContent,
+    PathGlobs,
+)
+from pants.engine.internals.native_engine import EMPTY_DIGEST, Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule, rule_helper
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class CGoCompileRequest(EngineAwareParameter):
+    import_path: str
+    pkg_name: str
+    digest: Digest
+    dir_path: str
+    cgo_files: tuple[str, ...]
+    cgo_flags: CGoCompilerFlags
+    cxx_files: tuple[str, ...] = ()
+    m_files: tuple[str, ...] = ()
+    f_files: tuple[str, ...] = ()
+
+    def debug_hint(self) -> str | None:
+        return self.import_path
+
+
+@dataclass(frozen=True)
+class CGoCompileResult:
+    digest: Digest
+    output_go_files: tuple[str, ...]
+    output_obj_files: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class CGoCompilerFlags:
+    cflags: tuple[str, ...]
+    cppflags: tuple[str, ...]
+    cxxflags: tuple[str, ...]
+    fflags: tuple[str, ...]
+    ldflags: tuple[str, ...]
+    pkg_config: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class CheckCompilerSupportsFlagRequest:
+    cc: str
+    flag: str
+
+
+@dataclass(frozen=True)
+class CheckCompilerSupportsOptionResult:
+    supports_flag: bool
+
+
+# Logic and comments in this rule come from `go` at:
+# https://github.com/golang/go/blob/7eaad60737bc507596c56cec4951b089596ccc9e/src/cmd/go/internal/work/exec.go#L2570
```

Comment:
Good grief! ;-)

Reply:
Looks like it's not just this rule but much of this file, including several other functions below, that are ported from exec.go? This comment should emphasize that, and/or each such function should be commented with a reminder of that fact.

Reply:
Yes, I've basically transcribed much the `go build` internal logic to work as Pants rules. The cgo logic is very particular so matching that logic will hopefully provide a consistent experience for cgo users if they use Pants.

Reply:
Will add comments.